### PR TITLE
PROJ-95 Offline Functionality

### PR DESCRIPTION
**Pull request by leroxyl**
**Added:**

- endpoints for offline signing and verification functionality
  -  `/${uuid}/offline`                     | [Offline Sealing Hashes (chained) - original data](#UPP-Offline-Sealing)  |
  - `/${uuid}/offline/hash`             | [Offline Sealing Hashes (chained) - hash](#UPP-Offline-Sealing)              |
  - `/${uuid}/anchor/offline`          | [Offline Sealing Hashes (no chain) - original data](#UPP-Offline-Sealing) |
  - `/${uuid}/anchor/offline/hash` | [Offline Sealing Hashes (no chain) - hash](#UPP-Offline-Sealing)             |
  - `/verify/offline`                        | [Offline Verification - original data](#UPP-Offline-Verification)                     |
  - `/verify/offline/hash`               | [Offline Verification - hash](#UPP-Offline-Verification)                                 |
- individually configurable HTTP request timeouts for requests to upstream servers
  - 	IdentityServiceTimeoutMs   int64    `json:"identityServiceTimeoutMs" envconfig:"IDENTITY_SERVICE_TIMEOUT_MS"`    // time limit for requests to the ubirch identity service in milliseconds
  - 	AuthServiceTimeoutMs       int64    `json:"authServiceTimeoutMs" envconfig:"AUTH_SERVICE_TIMEOUT_MS"`            // time limit for requests to the ubirch authentication service (niomon) in milliseconds
  - 	VerifyServiceTimeoutMs     int64    `json:"verifyServiceTimeoutMs" envconfig:"VERIFY_SERVICE_TIMEOUT_MS"`        // time limit for requests to the ubirch verification service in milliseconds
  - 	VerificationTimeoutMs      int64    `json:"verificationTimeoutMs" envconfig:"VERIFICATION_TIMEOUT_MS"`                   // time limit for repeated attempts to verify a hash at the ubirch verification service in milliseconds
  - if values are not set in the configuration, the following default values will be used 
    ```
    defaultIdentityServiceTimeoutMs = 10_000
    defaultAuthServiceTimeoutMs     = 2_000
    defaultVerifyServiceTimeoutMs   = 600
    defaultVerificationTimeoutMs    = 2_000
    ```
- test cases for signer and verifier

**Changed:**

- renamed configuration parameter `registerAuth` to `staticAuth`
  - 	StaticAuth                  string    `json:"staticAuth" envconfig:"STATIC_AUTH"`                                          // static auth token needed for identity registration, csr creation or key deactivation
- setting the static auth token is optional. Endpoints, which use the static auth token, have to be enabled explicitly in the configuration
  - 	EnableRegistrationEndpoint  bool    `json:"enableRegistrationEndpoint" envconfig:"ENABLE_REGISTRATION_ENDPOINT"`         // expose endpoint for identity registration
  - 	EnableCSRCreationEndpoint   bool    `json:"enableCSRCreationEndpoint" envconfig:"ENABLE_CSR_CREATION_ENDPOINT"`          // expose endpoint for CSR creation
  - 	EnableDeactivationEndpoint  bool    `json:"enableDeactivationEndpoint" envconfig:"ENABLE_DEACTIVATION_ENDPOINT"`         // expose endpoint for key status updates (de-/re-activation)

**Fixed:**

- use of HTTP connection pool for requests to upstream servers (http.DefaultClient)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -266,55 +297,109 @@ The "password" is the [UBIRCH backend authentication token](#how-to-acquire-the-
...
+| disable          | `/<UUID>/disable`    | `/<UUID>/disable/hash` |
+| enable           | `/<UUID>/enable`     | `/<UUID>/enable/hash`  |
+| delete           | `/<UUID>/delete`     | `/<UUID>/delete/hash`  |
 
 Hash update requests to the UBIRCH backend must come from the same UUID that anchored said hash and be signed by the
 same private key that signed the anchoring request.
 
+#### UPP Offline Sealing
+
+The client supports offline hash sealing, where the created UPP is not sent to the UBIRCH backend, but only retrieved in
```
Comment (ghost) [inline README.md]: 'only retrieved' should be 'instead only returned' I think.

File: main/adapters/clients/ubirch_identity_client.go [hunk 2]
```diff
@@ -90,11 +92,11 @@ func (c *IdentityServiceClient) IsKeyRegistered(id uuid.UUID, pubKey []byte) (bo
 }
 
 func (c *IdentityServiceClient) SubmitKeyRegistration(uid uuid.UUID, cert []byte) error {
-	log.Debugf("%s: registering public key at key service", uid)
+	log.Debugf("%s: registering public key at key service: %s", uid, cert)
```
Comment (ghost) [inline main/adapters/clients/ubirch_identity_client.go]: Optional: Maybe add 'with cert:' at the end to make it clearer that the cert is printed and not some service identifier.

File: main/adapters/handlers/signer_test.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+			name: "chain online",
+			msg: h.HTTPRequest{
+				Ctx:       context.Background(),
+				ID:        testUuid,
+				Auth:      testAuth,
+				Hash:      testHash,
+				Operation: h.ChainHash,
+				Offline:   false,
+			},
+			setExpectations: func(m *mock.Mock) {
```
Comment (ghost) [inline main/adapters/handlers/signer_test.go]: Might be more understandable to name this 'setMockBehavior'. But unsure if this violates conventions and its better left as-is.

File: main/adapters/http_server/csr_fetcher.go [hunk 4]
```diff
@@ -11,6 +11,9 @@ import (
 type GetCSR func(uid uuid.UUID) (csr []byte, err error)
 
 func FetchCSR(auth string, get GetCSR) http.HandlerFunc {
+	if len(auth) == 0 {
+		panic("missing auth token for CSR creation endpoint")
```
Comment (ghost) [inline main/adapters/http_server/csr_fetcher.go:15]: Is it really intended that we crash and burn here instead of giving an error?

File: main/adapters/http_server/http_server.go [hunk 5]
```diff
@@ -55,14 +165,24 @@ func (srv *HTTPServer) SetUpCORS(allowedOrigins []string, debug bool) {
 	}))
 }
 
-func (srv *HTTPServer) AddServiceEndpoint(endpoint ServerEndpoint) {
-	hashEndpointPath := path.Join(endpoint.Path, HashEndpoint)
+func HandleOptions(http.ResponseWriter, *http.Request) {}
+
+func (srv *HTTPServer) AddServiceEndpoint(endpointPath string, handle func(offline, isHash bool) http.HandlerFunc, supportOffline bool) {
```
Comment (ghost) [inline main/adapters/http_server/http_server.go:170]: It might be easier-to-read code to rename handle to handler or even handlerFunc to avoid mistaking it for a handle in the key/object handle sense.

File: main/adapters/http_server/http_server.go [hunk 6]
```diff
@@ -43,6 +42,117 @@ func NewRouter() *chi.Mux {
...
+	verificationService := &VerificationService{
+		Verify:        verify,
+		VerifyOffline: verifyOffline,
+	}
+
+	// verify:              /verify
+	// verify hash:         /verify/hash
+	// verify offline:      /verify/offline
+	// verify offline hash: /verify/offline/hash
+	httpServer.AddServiceEndpoint(VerifyPath,
```
Comment (ghost) [inline main/adapters/http_server/http_server.go:148]: The structure of adding the endpoints via multiple wrappers seems a bit convoluted to me. I was wondering if it's easier to set a list/struct with the endpoints and handlers first and the simply iterate over that. That might be more readable. But maybe there are also some other requirements I am not getting.

File: main/adapters/http_server/signing_service.go [hunk 7]
```diff
@@ -2,82 +2,80 @@ package http_server
...
-func getOperation(r *http.Request) (Operation, error) {
-	opParam := chi.URLParam(r, OperationKey)
-	switch Operation(opParam) {
-	case AnchorHash, DisableHash, EnableHash, DeleteHash:
-		return Operation(opParam), nil
-	default:
-		return "", fmt.Errorf("invalid operation: "+
-			"expected (\"%s\" | \"%s\" | \"%s\" | \"%s\"), got \"%s\"",
-			AnchorHash, DisableHash, EnableHash, DeleteHash, opParam)
+func (s *SigningService) HandleRequest(op Operation) func(bool, bool) http.HandlerFunc {
```
Comment (ghost) [inline main/adapters/http_server/signing_service.go]: It might be sensible to call the function-returning functions getRequestHandler or similar to make the code more readable.

File: main/adapters/http_server/signing_service.go [hunk 8]
```diff
@@ -2,82 +2,80 @@ package http_server
...
+	return func(offline, isHashRequest bool) http.HandlerFunc {
```
Comment (ghost) [inline main/adapters/http_server/signing_service.go:39]: The function handleRequest returns a function that returns another function which is hard to follow for me. Maybe a few comments here could help understand someone with less experience what is happening.

File: main/adapters/http_server/verification_service.go [hunk 9]
```diff
@@ -1,24 +1,68 @@
...
+type VerifyOffline func(upp []byte, hash []byte) HTTPResponse
+
 type VerificationService struct {
-	Verify func([]byte) HTTPResponse
+	Verify
+	VerifyOffline
 }
 
-var _ Service = (*VerificationService)(nil)
+func (s *VerificationService) HandleRequest(offline, isHashRequest bool) http.HandlerFunc {
```
Comment (ghost) [inline main/adapters/http_server/verification_service.go]: Same remark as in https://github.com/ubirch/ubirch-client-go/pull/102/files#diff-440fd7ece871c65eb098415fc61091366e9bd354bed520ce3d08206850b7c2c0R30

File: main/adapters/http_server/http_server_test.go [hunk 10]
```diff
@@ -2,11 +2,905 @@ package http_server
...
+					//Header:     http.Header{"test": []string{"1", "2", "3"}},
+					Header:  http.Header{"test": []string{"header"}},
+					Content: []byte("chained"),
+				})
+			},
+			tcChecks: func(t *testing.T, w *httptest.ResponseRecorder, m *mock.Mock) {
+				m.AssertExpectations(t)
+				assert.Equal(t, http.StatusOK, w.Code)
+				//assert.Equal(t, []string{"1", "2", "3"}, w.Header().Get("test"))
+				assert.Equal(t, "header", w.Header().Get("test"))
```
Comment (ghost) [inline main/adapters/http_server/http_server_test.go]: I don't get what this tests for.

File: main/adapters/http_server/http_server_test.go [hunk 11]
```diff
@@ -2,11 +2,905 @@ package http_server
...
+				).Return(HTTPResponse{
+					StatusCode: http.StatusOK,
+					Header:     http.Header{"test": []string{"header"}},
+					Content:    []byte("chained hash"),
+				})
+			},
+			tcChecks: func(t *testing.T, w *httptest.ResponseRecorder, m *mock.Mock) {
+				m.AssertExpectations(t)
+				assert.Equal(t, http.StatusOK, w.Code)
+				assert.Equal(t, "header", w.Header().Get("test"))
```
Comment (ghost) [inline main/adapters/http_server/http_server_test.go]: I don't get what this tests for. This happens a few more time, I will not make new comments for the other places.

File: README.md (hunk 1 above)
Comment (leroxyl) [reply README.md]: https://github.com/ubirch/ubirch-client-go/pull/102/commits/fe802e2332a09987c1346c88402794eb7707d261

File: main/adapters/http_server/csr_fetcher.go (hunk 4 above)
Comment (leroxyl) [reply main/adapters/http_server/csr_fetcher.go:15]: yes, as this function returns only a function and no error. With this check I wanted to make sure that it would be impossible to initialize this endpoint without an auth token.

File: main/adapters/http_server/http_server_test.go (hunk 10 above)
Comment (leroxyl) [reply main/adapters/http_server/http_server_test.go]: in this test, we call the InitHTTPServer function, which initializes all endpoints. Each test case sends a request to one endpoint to check if the request format will be interpreted as expected, the correct callbacks will be called  with the coreect parameters, and the response fits the expectations.

File: main/adapters/http_server/http_server_test.go (hunk 11 above)
Comment (leroxyl) [reply main/adapters/http_server/http_server_test.go]: same as above

File: main/adapters/clients/ubirch_identity_client.go (hunk 2 above)
Comment (leroxyl) [reply main/adapters/clients/ubirch_identity_client.go]: https://github.com/ubirch/ubirch-client-go/pull/102/commits/b4d9ba4c242b85adf2225322133f03ff1f2693ef

File: main/adapters/handlers/signer_test.go (hunk 3 above)
Comment (leroxyl) [reply main/adapters/handlers/signer_test.go]: https://github.com/ubirch/ubirch-client-go/pull/102/commits/987c2ac2400ff2e68ccd24742e7154150d51e10d

File: main/adapters/http_server/http_server.go (hunk 5 above)
Comment (leroxyl) [reply main/adapters/http_server/http_server.go:170]: i would rather leave it as it is because "handle" is a verb and I'm trying to follow the convention of naming functions with verbs

File: main/adapters/http_server/http_server.go (hunk 6 above)
Comment (leroxyl) [reply main/adapters/http_server/http_server.go:148]: you made a very good point here. This part became convoluted over time and is hard to read as well as to maintain. I took a look at it with Carlos and we decided that we should work on this, but not in the scope of this PR.

File: main/adapters/http_server/signing_service.go (hunk 8 above)
Comment (leroxyl) [reply main/adapters/http_server/signing_service.go:39]: Here is a first try: https://github.com/ubirch/ubirch-client-go/pull/102/commits/341883480a54c7bd5e345fdf9a6dde0f3d45d291

The size of the comment underlines the necessarily of a refactoring ;)

File: main/adapters/http_server/verification_service.go (hunk 9 above)
Comment (leroxyl) [reply main/adapters/http_server/verification_service.go]: I tried: https://github.com/ubirch/ubirch-client-go/pull/102/commits/5f22fe05f638f482fbb668e3edab843e64915007

File: main/adapters/http_server/http_server.go (hunk 5 above)
Comment (ghost) [reply main/adapters/http_server/http_server.go:170]: I am more used to encountering handle as a noun in programming context, thats probably also the reason for my confusion. I think I get it now and we can leave it as it is.

File: main/adapters/http_server/signing_service.go (hunk 8 above)
Comment (ghost) [reply main/adapters/http_server/signing_service.go:39]: Looks good. And I guess you are right this might be a good place to un-nest stuff when we find some time. ;)

File: main/adapters/http_server/http_server_test.go (hunk 10 above)
Comment (ghost) [reply main/adapters/http_server/http_server_test.go]: Sorry, my comment was a bit vague: what I did not understand was why we test for a header called 'test' that must contain the string 'header'. I have the feeling now this is simply a test of header setting/parsing. Am I correct? If so, the naming was simply a bit confusing for me. If it was a header called 'testheader' with a value of 'testvalue' I think I would have understood it more easily.

Comment (leroxyl) [reply main/adapters/http_server/http_server_test.go]: https://github.com/ubirch/ubirch-client-go/pull/102/commits/ac193f8bd475f9eee8ac56bca4952a415a26efe8
